File: go/internal/tools/linode_volumes.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/config"
	linodev1 "github.com/chadit/LinodeMCP/go/internal/genpb/linode/mcp/v1"
	"github.com/chadit/LinodeMCP/go/internal/linode"
	"github.com/chadit/LinodeMCP/go/internal/profiles"
	"github.com/chadit/LinodeMCP/go/internal/toolschemas"
)

// NewLinodeVolumeGetTool creates a tool for retrieving a single block storage volume by ID.
func NewLinodeVolumeGetTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool := mcp.NewToolWithRawSchema(
		"linode_volume_get",
		"Gets details for a single block storage volume by ID.",
		toolschemas.Schema("linode.mcp.v1.VolumeGetInput"),
	)

	handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		return handleLinodeVolumeGetRequest(ctx, &request, cfg)
	}

	return tool, profiles.CapRead, handler
}

func handleLinodeVolumeGetRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	volumeID := request.GetInt("volume_id", 0)
	if volumeID == 0 {
		return mcp.NewToolResultError("volume_id is required"), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	volume, err := client.GetVolumeProto(ctx, volumeID)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to retrieve volume %d: %v", volumeID, err)), nil
	}

	// Wrapped under a "volume" key to match the Python implementation's
	// response shape for this tool.
	return MarshalProtoToolResponse(&linodev1.VolumeGetResponse{Volume: volume})
}

// NewLinodeVolumeListTool creates a tool for listing Linode block storage volumes.
func NewLinodeVolumeListTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newProtoListToolRawSchema(
		cfg,
		"linode_volume_list",
		"Lists all block storage volumes for the authenticated user with optional filtering by region or label",
		"linode.mcp.v1.VolumeListInput",
		func(ctx context.Context, client *linode.Client) ([]*linodev1.Volume, error) {
			return client.ListVolumesProto(ctx)
		},
		[]listFilterParam[*linodev1.Volume]{
			fieldFilter("region", "Filter volumes by region (e.g., 'us-east', 'eu-west')",
				func(vol *linodev1.Volume) string { return vol.GetRegion() }),
			containsFilter("label_contains", "Filter volumes where label contains this string (case-insensitive)",
				func(vol *linodev1.Volume) string { return vol.GetLabel() }),
		},
		volumeListResponse,
	)

	return tool, profiles.CapRead, handler
}

func volumeListResponse(items []*linodev1.Volume, count int32, filter *string) *linodev1.VolumeListResponse {
	return &linodev1.VolumeListResponse{Count: count, Filter: filter, Volumes: items}
}

// NewLinodeVolumeTypeListTool creates a tool for listing Linode block storage volume types.
func NewLinodeVolumeTypeListTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newProtoListToolRawSchema(
		cfg,
		"linode_volume_type_list",
		"Lists available Linode block storage volume types and pricing",
		"linode.mcp.v1.VolumeTypeListInput",
		func(ctx context.Context, client *linode.Client) ([]*linodev1.LinodeType, error) {
			return client.ListVolumeTypesProto(ctx)
		},
		nil,
		volumeTypeListResponse,
	)

	return tool, profiles.CapRead, handler
}

func volumeTypeListResponse(items []*linodev1.LinodeType, count int32, filter *string) *linodev1.VolumeTypeListResponse {
	return &linodev1.VolumeTypeListResponse{Count: count, Filter: filter, VolumeTypes: items}
}

File: python/src/linodemcp/tools/linode_volumes.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from mcp.types import TextContent, Tool

from linodemcp.genpb.linode.mcp.v1 import type_pb2, volume_pb2
from linodemcp.profiles import Capability
from linodemcp.tools.helpers import execute_tool
from linodemcp.tools.proto_response import (
    serialize_api_response,
    serialize_list_response,
)
from linodemcp.tools.toolschemas import schema

if TYPE_CHECKING:
    from linodemcp.config import Config
    from linodemcp.linode import RetryableClient


def create_linode_volume_get_tool() -> tuple[Tool, Capability]:
    """Create the linode_volume_get tool."""
    return Tool(
        name="linode_volume_get",
        description="Gets details for a single block storage volume by ID.",
        inputSchema=schema("linode.mcp.v1.VolumeGetInput"),
    ), Capability.Read


async def handle_linode_volume_get(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_volume_get tool request."""
    volume_id = arguments.get("volume_id", 0)
    if not volume_id:
        return [TextContent(type="text", text="Error: volume_id is required")]

    async def _call(client: RetryableClient) -> dict[str, Any]:
        raw = await client.get_raw(f"/volumes/{int(volume_id)}")
        return serialize_api_response({"volume": raw}, volume_pb2.VolumeGetResponse())

    return await execute_tool(cfg, arguments, "retrieve Linode volume", _call)


def create_linode_volume_list_tool() -> tuple[Tool, Capability]:
    """Create the linode_volume_list tool."""
    return Tool(
        name="linode_volume_list",
        description=(
            "Lists all block storage volumes for the authenticated user "
            "with optional filtering by region or label"
        ),
        inputSchema=schema("linode.mcp.v1.VolumeListInput"),
    ), Capability.Read


def create_linode_volume_type_list_tool() -> tuple[Tool, Capability]:
    """Create the linode_volume_type_list tool."""
    return Tool(
        name="linode_volume_type_list",
        description="Lists available block storage volume types and prices.",
        inputSchema=schema("linode.mcp.v1.VolumeTypeListInput"),
    ), Capability.Read


async def handle_linode_volume_type_list(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_volume_type_list tool request."""

    async def _call(client: RetryableClient) -> dict[str, Any]:
        volume_types = await client.list_volume_types()
        return serialize_list_response(
            {"data": volume_types},
            "volume_types",
            type_pb2.VolumeTypeListResponse(),
        )

    return await execute_tool(cfg, arguments, "list Linode volume types", _call)


async def handle_linode_volume_list(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_volume_list tool request."""
    region_filter: str = arguments.get("region", "")
    label_contains: str = arguments.get("label_contains", "")

    def _matches(volume: dict[str, Any]) -> bool:
        region = str(volume.get("region", ""))
        if region_filter and region.lower() != region_filter.lower():
            return False
        label = str(volume.get("label", ""))
        return not (label_contains and label_contains.lower() not in label.lower())

    filters: list[str] = []
    if region_filter:
        filters.append(f"region={region_filter}")
    if label_contains:
        filters.append(f"label_contains={label_contains}")

    async def _call(client: RetryableClient) -> dict[str, Any]:
        raw = await client.get_raw("/volumes")
        return serialize_list_response(
            raw,
            "volumes",
            volume_pb2.VolumeListResponse(),
            filter_value=", ".join(filters) if filters else None,
            item_filter=_matches,
        )

    return await execute_tool(cfg, arguments, "retrieve Linode volumes", _call)
